internal/handlers/common/aggregations/collstats.go

```diff
@@ -0,0 +1,112 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package aggregations
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/FerretDB/FerretDB/internal/handlers/common"
+	"github.com/FerretDB/FerretDB/internal/handlers/commonerrors"
+	"github.com/FerretDB/FerretDB/internal/types"
+	"github.com/FerretDB/FerretDB/internal/util/must"
+)
+
+// collStats represents $collStats stage.
+type collStats struct {
+	storageStats   *storageStats
+	count          bool
+	latencyStats   bool
+	queryExecStats bool
+}
+
+// storageStats represents $collStats.storageStats field.
+type storageStats struct {
+	scale int32
+}
+
+// newCollStats creates a new $collStats stage.
+func newCollStats(stage *types.Document) (Stage, error) {
+	fields, err := common.GetRequiredParam[*types.Document](stage, "$collStats")
+	if err != nil {
+		return nil, commonerrors.NewCommandErrorMsgWithArgument(
+			commonerrors.ErrStageCollStatsInvalidArg,
+			fmt.Sprintf("$collStats must take a nested object but found: %s", types.FormatAnyValue(stage)),
+			"$collStats (stage)",
+		)
+	}
+
+	var cs collStats
+
+	// TODO Return error on invalid type of count: https://github.com/FerretDB/FerretDB/issues/2336
+	cs.count = fields.Has("count")
+
+	// TODO Implement latencyStats: https://github.com/FerretDB/FerretDB/issues/2341
+	cs.latencyStats = fields.Has("latencyStats")
+
+	// TODO Implement queryExecStats: https://github.com/FerretDB/FerretDB/issues/2341
+	cs.queryExecStats = fields.Has("queryExecStats")
+
+	if fields.Has("storageStats") {
+		cs.storageStats = new(storageStats)
+
+		// TODO Add proper support for scale: https://github.com/FerretDB/FerretDB/issues/1346
+		cs.storageStats.scale, err = common.GetOptionalPositiveNumber(
+			must.NotFail(fields.Get("storageStats")).(*types.Document),
+			"scale",
+		)
+		if err != nil || cs.storageStats.scale == 0 {
+			cs.storageStats.scale = 1
+		}
+	}
+
+	return &cs, nil
+}
+
+// Process implements Stage interface.
+//
+// Processing consists of modification of the input document, so it contains all the necessary fields
+// and the data is modified according to the given request.
+func (c *collStats) Process(ctx context.Context, in []*types.Document) ([]*types.Document, error) {
+	// The result of $collStats stage is always an array with a single document, and we expect the same input.
+	if len(in) != 1 {
+		panic(fmt.Sprintf("collStats: expected 1 document, got %d", len(in)))
+	}
+
+	res := in[0]
+
+	if c.storageStats != nil && c.storageStats.scale > 1 {
+		scale := float64(c.storageStats.scale)
+
+		scalable := []string{"size", "storageSize", "freeStorageSize", "totalIndexSize"}
+		for _, key := range scalable {
+			if res.Has(key) {
```

Comment:
I think it has to be next layer.
```suggestion
storageStat := must.NotFail(res.Get("storageStats"))
			if storageStat.Has(key) {
```

Currently scale isn't applied.
```
test> db.test.aggregate([{$collStats:{count:{}, storageStats:{scale:1000}}}])
[
  {
    ns: 'test.test',
    host: 'alma',
    localTime: '2023-04-03T02:24:57Z',
    storageStats: {
      size: Long("32768"),
      count: 2,
      avgObjSize: 4096,
      storageSize: Long("8192"),
      freeStorageSize: 0,
      capped: false,
      wiredTiger: {},
      nindexes: 1,
      indexDetails: {},
      indexBuilds: {},
      totalIndexSize: Long("16384"),
      indexSizes: {}
    },
    count: 2
  }
]
test> db.test.aggregate([{$collStats:{count:{}, storageStats:{}}}])
[
  {
    ns: 'test.test',
    host: 'alma',
    localTime: '2023-04-03T02:25:05Z',
    storageStats: {
      size: Long("32768"),
      count: 2,
      avgObjSize: 4096,
      storageSize: Long("8192"),
      freeStorageSize: 0,
      capped: false,
      wiredTiger: {},
      nindexes: 1,
      indexDetails: {},
      indexBuilds: {},
      totalIndexSize: Long("16384"),
      indexSizes: {}
    },
    count: 2
  }
]
```

Reply:
Nice catch!

Fixes are pushed:

```
test> db.values.aggregate( [ { $collStats: { storageStats: {scale: 1} } } ] )
[
  {
    ns: 'test.values',
    host: 'ferretdb',
    localTime: '2023-04-03T08:33:38Z',
    storageStats: {
      size: 65536,
      count: 69,
      avgObjSize: 237.44927536231884,
      storageSize: 16384,
      freeStorageSize: 0,
      capped: false,
      wiredTiger: {},
      nindexes: 1,
      indexDetails: {},
      indexBuilds: {},
      totalIndexSize: 16384,
      indexSizes: {}
    }
  }
]
test> db.values.aggregate( [ { $collStats: { storageStats: {scale: 1000} } } ] )
[
  {
    ns: 'test.values',
    host: 'ferretdb',
    localTime: '2023-04-03T08:33:43Z',
    storageStats: {
      size: 65.536,
      count: 69,
      avgObjSize: 237.44927536231884,
      storageSize: 16.384,
      freeStorageSize: 0,
      capped: false,
      wiredTiger: {},
      nindexes: 1,
      indexDetails: {},
      indexBuilds: {},
      totalIndexSize: 16.384,
      indexSizes: {}
    }
  }
]```

Reply:
Actually, now I wonder if `avgObjSize` should also use `scale` 🤔 I'll check what proxy returns!

Reply:
Fixed